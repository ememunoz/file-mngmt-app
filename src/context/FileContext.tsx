import AsyncStorage from '@react-native-async-storage/async-storage'
import {
  ReactNode,
  createContext,
  useContext,
  useEffect,
  useState,
} from 'react'

type ImageInfoType = {
  src: string
}

type SheetInfoType = {
  src: string
  total: number
  name: string
}

type FileStorageType = {
  imageInfoArray: ImageInfoType[]
  sheetInfoArray: SheetInfoType[]
}

type FileContextType = {
  imageInfoArray: ImageInfoType[]
  sheetInfoArray: SheetInfoType[]
  addImage: (src: string) => void
  addSheet: (obj: SheetInfoType) => void
}

const FileContext = createContext<FileContextType>({
  imageInfoArray: [],
  sheetInfoArray: [],
  addImage: () => {},
  addSheet: () => {},
})

const getInitialFileContext = async () => {
  const defaultValue: FileStorageType = {
    imageInfoArray: [],
    sheetInfoArray: [],
  }
  try {
    const rawValue = await AsyncStorage.getItem('@file-mngmt-state')
    const jsonValue: FileStorageType = !!rawValue
      ? (JSON.parse(rawValue) as FileStorageType)
      : defaultValue
    return jsonValue
  } catch (e) {
    return defaultValue
  }
}

export const FileContextProvider = ({ children }: { children: ReactNode }) => {
  const [imageInfoArray, setImageInfoArray] = useState<ImageInfoType[]>([])
  const [sheetInfoArray, setSheetInfoArray] = useState<SheetInfoType[]>([])

  useEffect(() => {
    const getDefaultValues = async () => {
      const { imageInfoArray, sheetInfoArray } = await getInitialFileContext()
      setImageInfoArray(imageInfoArray)
      setSheetInfoArray(sheetInfoArray)
    }
    getDefaultValues()
  }, [])

  const updateStorage = async () => {
    const data: FileStorageType = {
      imageInfoArray,
      sheetInfoArray,
    }
    await AsyncStorage.setItem('@file-mngmt-state', JSON.stringify(data))
  }

  const addImage = (src: string) => {
    const newArray = [...imageInfoArray, { src }]
    setImageInfoArray(newArray)
    updateStorage()
  }

  const addSheet = ({ src, name, total }: SheetInfoType) => {
    const newArray = [...sheetInfoArray, { src, name, total }]
    setSheetInfoArray(newArray)
    updateStorage()
  }

  return (
    <FileContext.Provider
      value={{ imageInfoArray, sheetInfoArray, addImage, addSheet }}
    >
      {children}
    </FileContext.Provider>
  )
}

export const useFileContext = () => useContext(FileContext)
