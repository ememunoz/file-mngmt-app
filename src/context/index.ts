export { FileContextProvider, useFileContext } from './FileContext/FileContext'
