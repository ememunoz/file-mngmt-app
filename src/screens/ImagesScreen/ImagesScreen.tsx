import { FlatList, Image, StyleSheet, Text, View } from 'react-native'

import { useFileContext } from '../../context'

const ImagesScreen = () => {
  const { imageInfoArray } = useFileContext()

  const renderItem = ({ item }) => (
    <Image
      source={{ uri: item.src }}
      style={styles.gallery}
      resizeMode="cover"
    />
  )

  return (
    <View style={styles.container}>
      {imageInfoArray.length > 0 ? (
        <FlatList
          data={imageInfoArray}
          numColumns={2}
          keyExtractor={item => item.src}
          renderItem={renderItem}
        />
      ) : (
        <Text style={styles.message}>
          You haven't uploaded any image, you can upload them in home page! ðŸ˜„
        </Text>
      )}
    </View>
  )
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    marginVertical: 16,
    marginHorizontal: 2,
    justifyContent: 'center',
  },
  gallery: {
    aspectRatio: 1,
    flex: 1 / 2,
    marginHorizontal: 2,
    marginVertical: 2,
  },
  message: {
    alignSelf: 'center',
    fontSize: 24,
    color: 'hsl(0, 0%, 60%)',
    textAlign: 'center',
    marginHorizontal: 16,
  },
})

export default ImagesScreen
