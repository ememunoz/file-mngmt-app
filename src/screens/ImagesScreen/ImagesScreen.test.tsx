import { render, screen } from '@testing-library/react-native'
import React from 'react'

import { FileContextProvider } from '../../context'
import ImagesScreen from './ImagesScreen'

describe('<ImagesScreen />', () => {
  it('renders correctly', () => {
    const component = (
      <FileContextProvider>
        <ImagesScreen />
      </FileContextProvider>
    )
    render(component)
    expect(screen.toJSON()).toMatchSnapshot()
  })
})
