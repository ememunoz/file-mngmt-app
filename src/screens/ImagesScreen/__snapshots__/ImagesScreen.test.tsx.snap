// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ImagesScreen /> renders correctly 1`] = `
<View
  style={
    Object {
      "flex": 1,
      "justifyContent": "center",
      "marginHorizontal": 2,
      "marginVertical": 16,
    }
  }
>
  <Text
    style={
      Object {
        "alignSelf": "center",
        "color": "hsl(0, 0%, 60%)",
        "fontSize": 24,
        "marginHorizontal": 16,
        "textAlign": "center",
      }
    }
  >
    You haven't uploaded any image, you can upload them in home page! ðŸ˜„
  </Text>
</View>
`;
